caproto/_log.py

```diff
@@ -143,9 +151,169 @@ def color_logs(color):
+ch_logger = logging.getLogger('caproto.ch')
+search_logger = logging.getLogger('caproto.bcast.search')
 current_handler = None  # overwritten below
 
 
+CRITICAL = 50
+FATAL = CRITICAL
+ERROR = 40
+WARNING = 30
+WARN = WARNING
+INFO = 20
+DEBUG = 10
+NOTSET = 0
+
+
+_levelToName = {
+    CRITICAL: 'CRITICAL',
+    ERROR: 'ERROR',
+    WARNING: 'WARNING',
+    INFO: 'INFO',
+    DEBUG: 'DEBUG',
+    NOTSET: 'NOTSET',
+}
+
+
+_nameToLevel = {
+    'CRITICAL': CRITICAL,
+    'FATAL': FATAL,
+    'ERROR': ERROR,
+    'WARN': WARNING,
+    'WARNING': WARNING,
+    'INFO': INFO,
+    'DEBUG': DEBUG,
+    'NOTSET': NOTSET,
+}
+
+
+class logger_nameFilter(logging.Filter):
+    def __init__(self, logger_names):
+        self.logger_names = logger_names
+    def filter(self, record):
+        for i in self.logger_names:
+            if fnmatch.fnmatch(record.name, i):
+                return True
+        return False
+
+
+class PVFilter(logging.Filter):
+    '''
+    Pass through only messages relevant to one or more PV names and
+    pv-unrelated messages.
+
+    Parameters
+    ----------
+    names : string or list of string
+        PV name or PV name list which will be filtered in.
+
+    Returns
+    -------
+    Bool: True or False
+        True if message is not PV related which doesn't have 'pv' as key in extra
+        True if 'pv' as key exists and pv name exists in Filter list.
+        False if message is PV related but pv name isn't in Filter list.
+    '''
+    def __init__(self, names, level='NOTSET', pv_unrelated_flag=True):
+        self.names = names
+        self.level = level
+        self.pv_unrelated_flag = pv_unrelated_flag
+
+    def filter(self, record):
+        if hasattr(record, 'pv'):
+            for i in self.names:
+                if fnmatch.fnmatch(record.pv, i) and record.levelno >= _nameToLevel[self.level]:
+                    return True
+            return False
+        else:
+            return self.pv_unrelated_flag
+
+
+class PVOnlyFilter(logging.Filter):
+    '''
+    99% as same as PVFilter class but only default return for PV unrelated message
+    Pass through only messages relevant to one or more PV names.
+
+    Parameters
+    ----------
+    names : string or list of string
+        PV name or PV name list which will be filtered in.
+
+    Returns
+    -------
+    Bool: True or False
+        False if message is not PV related which doesn't have 'pv' as key in extra
+        True if 'pv' as key exists and pv name exists in Filter list.
+        False if message is PV related but pv name isn't in Filter list.
+    '''
+    def __init__(self, names, level='NOTSET', pv_unrelated_flag=False):
+        self.names = names
+        self.level = level
+        self.pv_unrelated_flag = pv_unrelated_flag
+
+    def filter(self, record):
+        if hasattr(record, 'pv'):
+            for i in self.names:
+                if fnmatch.fnmatch(record.pv, i) and record.levelno >= _nameToLevel[self.level]:
```

Comment:
Good idea making sure to convert the string to an integer level that we can compare. Python provides a function for this already, and it converts (somewhat confusingly) in both directions:

```py
>>> logging.getLevelName('DEBUG')
10
>>> logging.getLevelName(10)
'DEBUG'
```

More improtantly, I think we can do this conversion just once in `__init__` before we set `self.level` so that we don't pay the cost of doing it again every time a new `record` flows through the filter. Agree?

Reply:
Sure